projects/vdk-plugins/vdk-jupyter/vdk-jupyterlab-extension/vdk_jupyterlab_extension/oauth2.py

```diff
@@ -0,0 +1,136 @@
+        # But that module is written with focus on CLI usage a bit making it harder to reuse
+        # and it needs to be refactored first.
+        redirect_url = self._redirect_url
+        if not redirect_url:
+            redirect_url = self.request.full_url()
+        redirect_url = self._fix_localhost(redirect_url)
+
+        log.info(f"redirect uri is {redirect_url}")
+
+        if self.get_argument("code", None):
+            log.info(
+                "Authorization code received. Will generate access token using authorization code."
+            )
+            tokens = self._exchange_auth_code_for_access_token(redirect_url)
+            log.info(f"Got tokens data: {tokens}")  # TODO: remove this
+            self._persist_tokens_data(tokens)
+        else:
+            log.info(f"Authorization URL is: {self._authorization_url}")
+            full_authorization_url = self._prepare_authorization_code_request_url(
+                redirect_url
+            )
+            self.finish(full_authorization_url)
+
+    def _persist_tokens_data(self, tokens):
+        auth = BaseAuth(conf=InMemAuthConfiguration())
+        auth.update_oauth2_authorization_url(self._access_token_url)
+        auth.update_client_id(self._client_id)
+        auth.update_access_token(tokens.get(AuthRequestValues.ACCESS_TOKEN_KEY.value))
+        auth.update_access_token_expiration_time(
+            time.time() + int(tokens[AuthRequestValues.EXPIRATION_TIME_KEY.value])
+        )
+        if AuthRequestValues.REFRESH_TOKEN_GRANT_TYPE in tokens:
+            auth.update_refresh_token(
+                tokens.get(AuthRequestValues.REFRESH_TOKEN_GRANT_TYPE)
+            )
+
+    def _prepare_authorization_code_request_url(self, redirect_uri):
+        (code_verifier, code_challenge, code_challenge_method) = generate_pkce_codes()
+        self.application.settings["code_verifier"] = code_verifier
+        oauth = OAuth2Session(client_id=self._client_id, redirect_uri=redirect_uri)
+        full_authorization_url = oauth.authorization_url(
+            self._authorization_url,
+            state="requested",
+            prompt=AuthRequestValues.LOGIN_PROMPT.value,
+            code_challenge=code_challenge,
+            code_challenge_method=code_challenge_method,
+        )[0]
+        return full_authorization_url
+
+    def _exchange_auth_code_for_access_token(self, redirect_uri) -> dict:
+        code = self.get_argument("code")
+        headers = {
+            AuthRequestValues.CONTENT_TYPE_HEADER.value: AuthRequestValues.CONTENT_TYPE_URLENCODED.value,
+        }
+        code_verifier = self.application.settings["code_verifier"]
+
+        data = (
+            f"code={code}&"
+            + f"grant_type=authorization_code&"
+            + f"code_verifier={code_verifier}&"
+            f"redirect_uri={redirect_uri}"
+        )
+        basic_auth = HTTPBasicAuth(self._client_id, "")
+        try:
+            # TODO : this should be async io
```

Comment:
Why?

Reply:
It's ok to block if you cannot proceed with other operations without authentication. This is not the case here. Only few VDK-specific operations require this authentication. But as it's generally single user server it's probably ok to leave .